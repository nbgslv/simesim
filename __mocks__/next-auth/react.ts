const mockSession = {
  expires: new Date(Date.now() + 2 * 86400).toISOString(),
  user: {
    email: '0512345678',
    emailEmail: 'test@test.com',
    id: 'clcclcclcclcclcclc1234',
    image: null,
    name: 'foo bar',
    role: 'ADMIN',
  },
};

export const useSession = jest.fn(
  () => ({ data: mockSession, status: 'authenticated' }) // return type is [] in v3 but changed to {} in v4
);
export const signOut = jest.fn(() =>
  Promise.resolve({ url: process.env.NEXTAUTH_URL as string })
);
