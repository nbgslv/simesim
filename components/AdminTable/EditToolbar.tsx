import React from 'react';
import {
    GridRowId,
    GridRowModel,
    GridRowModes,
    GridRowModesModel,
    GridRowsProp,
    GridToolbarContainer, GridValidRowModel
} from '@mui/x-data-grid';
import {Button, Spinner} from "react-bootstrap";
import styles from './EditToolbar.module.scss'
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import { solid } from "@fortawesome/fontawesome-svg-core/import.macro";
import ConfirmationModal from "./ConfirmationModal";

interface EditToolbarProps<T extends GridValidRowModel> {
    selectedRows: GridRowId[]
    setRows: (newRows: (oldRows: GridRowsProp) => GridRowsProp) => void;
    setRowModesModel: (
        newModel: (oldModel: GridRowModesModel) => GridRowModesModel,
    ) => void;
    multiActions?: string[]
    addRow?: () => Promise<{ id: GridRowId, columnToFocus: string | undefined }>
    deleteRows?: (ids: GridRowId[]) => Promise<void>
    loading: boolean
}

const EditToolbar =
    <T extends GridValidRowModel>
    ({
        selectedRows,
        setRowModesModel,
        multiActions = [],
        addRow,
        deleteRows
    }: EditToolbarProps<T>) => {
        const [loading, setLoading] = React.useState<boolean>(false);
        const [showDeleteModal, setShowDeleteModal] = React.useState<boolean>(false);

        const handleAdd = async () => {
            try {
                setLoading(true)
                if (addRow) {
                    const newRowData = await addRow()
                    if (newRowData) {
                        setRowModesModel((oldModel) => ({
                            ...oldModel,
                            [newRowData.id]: { mode: GridRowModes.Edit, fieldToFocus: newRowData.columnToFocus },
                        }));
                    }
                }
            } finally {
                setLoading(false)
            }
        };

        const handleDelete = async () => {
            setShowDeleteModal(false)
            setLoading(true)
            await deleteRows?.(selectedRows)
            setLoading(false)
        }

        return (
            <div dir="rtl" className="position-relative">
                {loading &&
                    <div className={styles.spinner}>
                        <Spinner animation={'border'} />
                    </div>
                }
                <GridToolbarContainer>
                    {multiActions?.includes('add') &&
                      <Button className={styles.button} onClick={handleAdd}>
                        <FontAwesomeIcon icon={solid('plus')}/>
                          חדש
                      </Button>
                    }
                    {multiActions?.includes('delete') &&
                      <Button className={styles.button} disabled={!selectedRows.length} onClick={() => setShowDeleteModal(true)}>
                        <FontAwesomeIcon icon={solid('trash-can')}/>
                        מחק
                      </Button>
                    }
                </GridToolbarContainer>
                <ConfirmationModal
                    show={showDeleteModal}
                    title={'מחיקה'}
                    body={'בטוח שברצונך למחוק?'}
                    confirmButtonText={'כן'}
                    cancelButtonText={'לא'}
                    confirmAction={handleDelete}
                    cancelAction={() => setShowDeleteModal(false)}
                />
            </div>
        );
}

export default EditToolbar;
