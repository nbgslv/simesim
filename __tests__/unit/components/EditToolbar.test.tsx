import { render, fireEvent, waitFor } from '@testing-library/react';
import React from 'react';
import EditToolbar from '../../../components/AdminTable/EditToolbar';

jest.mock('@mui/x-data-grid', () => ({
  GridToolbarContainer: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
}));

describe('EditToolbar', () => {
  const props = {
    selectedRows: [],
    setRows: jest.fn(),
    setRowModesModel: jest.fn(),
    multiActions: ['add', 'delete'],
    addRow: jest.fn().mockResolvedValue({ id: 1, columnToFocus: 'name' }),
    deleteRows: jest.fn(),
    loading: false,
  };

  it('should render correctly', () => {
    const { getByText } = render(<EditToolbar {...props} />);
    expect(getByText('חדש')).toBeInTheDocument();
    expect(getByText('מחק')).toBeInTheDocument();
  });

  it('should call addRow when add button is clicked', async () => {
    const { getByText } = render(<EditToolbar {...props} />);
    fireEvent.click(getByText('חדש'));
    await waitFor(() => expect(props.addRow).toHaveBeenCalled());
    expect(props.setRowModesModel).toHaveBeenCalledWith(expect.any(Function));
    return expect(props.addRow()).resolves.toEqual({
      id: 1,
      columnToFocus: 'name',
    });
  });

  it.skip('should show delete confirmation modal when delete button is clicked', async () => {
    /*
    react-bootstrap's Modal component is not working with react-testing-library
    as it is being rendered into a portal.
    https://github.com/testing-library/react-testing-library/issues/62
    TODO: test when this option is available.
     */
  });

  it.skip('should call deleteRows when delete confirmation modal is confirmed', async () => {
    // TODO: test when react-bootstrap's Modal component is working with react-testing-library
  });
});
