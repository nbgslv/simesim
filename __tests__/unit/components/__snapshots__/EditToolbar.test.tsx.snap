// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditToolbar matches snapshot 1`] = `
<div>
  <div
    class="position-relative"
    dir="rtl"
  >
    <div>
      <button
        class="button btn btn-primary"
        type="button"
      >
        חדש
      </button>
      <button
        class="button btn btn-primary"
        disabled=""
        type="button"
      >
        מחק
      </button>
    </div>
  </div>
</div>
`;
